source/lac/trilinos_epetra_vector.cc

```diff
@@ -0,0 +1,529 @@
+      // Check that casting will work.
+      Assert(dynamic_cast<const Vector *>(&V)!=NULL,
+             ExcVectorTypeNotCompatible());
+
+      // Downcast V. If fails, throws an exception.
+      const Vector &down_V = dynamic_cast<const Vector &>(V);
+      // If we don't have the same map, copy.
+      if (vector->Map().SameAs(down_V.trilinos_vector().Map())==false)
+        this->sadd(0., a, V);
+      else
+        {
+          // Otherwise, just update
+          int ierr = vector->Update(a, down_V.trilinos_vector(), 0.);
+          Assert(ierr==0, ExcTrilinosError(ierr));
+        }
+    }
+
+
+
+    double Vector::l1_norm() const
+    {
+      double norm(0.);
+      int ierr = vector->Norm1(&norm);
+      Assert(ierr==0, ExcTrilinosError(ierr));
+
+      return norm;
+    }
+
+
+
+    double Vector::l2_norm() const
+    {
+      double norm(0.);
+      int ierr = vector->Norm2(&norm);
+      Assert(ierr==0, ExcTrilinosError(ierr));
+
+      return norm;
+    }
+
+
+
+    double Vector::linfty_norm() const
+    {
+      double norm(0.);
+      int ierr = vector->NormInf(&norm);
+      Assert(ierr==0, ExcTrilinosError(ierr));
+
+      return norm;
+    }
+
+
+
+    double Vector::add_and_dot(const double a,
+                               const VectorSpaceVector<double> &V,
+                               const VectorSpaceVector<double> &W)
+    {
+      this->add(a, V);
+
+      return *this * W;
+    }
+
+
+
+    Vector::size_type Vector::size() const
+    {
+#ifndef DEAL_II_WITH_64BIT_INDICES
+      return vector->GlobalLength();
+#else
+      return vector->GlobalLength64();
+#endif
+    }
+
+
+
+    MPI_Comm Vector::get_mpi_communicator() const
+    {
+      const Epetra_MpiComm *epetra_comm
+        = dynamic_cast<const Epetra_MpiComm *>(&(vector->Comm()));
+      return epetra_comm->GetMpiComm();
+    }
+
+
+
+    const ::dealii::IndexSet Vector::locally_owned_elements() const
+    {
+      const Epetra_Map *map = dynamic_cast<const Epetra_Map *>(&(vector->Map()));
+      return IndexSet(*map);
+    }
+
+
+
+    const Epetra_FEVector &Vector::trilinos_vector() const
+    {
+      return *vector;
+    }
+
+
+
+    Epetra_FEVector &Vector::trilinos_vector()
+    {
+      return *vector;
+    }
+
+
+
+    void Vector::print(std::ostream &out,
+                       const unsigned int precision,
+                       const bool scientific,
+                       const bool across) const
+    {
+      AssertThrow(out, ExcIO());
+
+      // Get a representation of the vector and loop over all
+      // the elements
+      double *val;
+      int leading_dimension;
+      int ierr = vector->ExtractView(&val, &leading_dimension);
+
+      Assert(ierr==0, ExcTrilinosError(ierr));
+      out.precision (precision);
+      if (scientific)
+        out.setf(std::ios::scientific, std::ios::floatfield);
+      else
+        out.setf(std::ios::fixed, std::ios::floatfield);
+
+      if (across)
+        for (size_type i=0; i<size(); ++i)
```

Comment:
I think you want to use `vector->MyLength()` here. `size()` is the global vector size but val is the locally owned array.